Mathlib/AlgebraicGeometry/EllipticCurve/DivisionPolynomial/Basic.lean

```diff
@@ -304,43 +319,60 @@ lemma Ψ_ofNat (n : ℕ) : W.Ψ n = C (W.preΨ' n) * if Even n then W.ψ₂ else
 
 @[simp]
 lemma Ψ_zero : W.Ψ 0 = 0 := by
-  erw [Ψ_ofNat, preΨ'_zero, C_0, zero_mul]
+  rw [← Nat.cast_zero, Ψ_ofNat, preΨ'_zero, C_0, zero_mul]
 
 @[simp]
 lemma Ψ_one : W.Ψ 1 = 1 := by
-  erw [Ψ_ofNat, preΨ'_one, C_1, mul_one]
+  rw [← Nat.cast_one, Ψ_ofNat, preΨ'_one, C_1, if_neg Nat.not_even_one, mul_one]
 
 @[simp]
 lemma Ψ_two : W.Ψ 2 = W.ψ₂ := by
-  erw [Ψ_ofNat, preΨ'_two, one_mul, if_pos even_two]
+  rw [← Nat.cast_two, Ψ_ofNat, preΨ'_two, C_1, one_mul, if_pos even_two]
 
 @[simp]
 lemma Ψ_three : W.Ψ 3 = C W.Ψ₃ := by
-  erw [Ψ_ofNat, preΨ'_three, mul_one]
+  rw [← Nat.cast_three, Ψ_ofNat, preΨ'_three, if_neg <| by decide, mul_one]
 
 @[simp]
 lemma Ψ_four : W.Ψ 4 = C W.preΨ₄ * W.ψ₂ := by
-  erw [Ψ_ofNat, preΨ'_four, if_pos <| by decide]
+  rw [← Nat.cast_four, Ψ_ofNat, preΨ'_four, if_pos <| by decide]
 
-lemma Ψ_odd (m : ℕ) : W.Ψ (2 * (m + 2) + 1) =
-    W.Ψ (m + 4) * W.Ψ (m + 2) ^ 3 - W.Ψ (m + 1) * W.Ψ (m + 3) ^ 3 +
-      W.toAffine.polynomial * (16 * W.toAffine.polynomial - 8 * W.ψ₂ ^ 2) * C
-        (if Even m then W.preΨ' (m + 4) * W.preΨ' (m + 2) ^ 3
-          else -W.preΨ' (m + 1) * W.preΨ' (m + 3) ^ 3) := by
-  repeat erw [Ψ_ofNat]
-  simp_rw [preΨ'_odd, if_neg (m + 2).not_even_two_mul_add_one, Nat.even_add_one, ite_not]
-  split_ifs <;> C_simp <;> rw [C_Ψ₂Sq] <;> ring1
-
-lemma Ψ_even (m : ℕ) : W.Ψ (2 * (m + 3)) * W.ψ₂ =
+lemma Ψ_even_ofNat (m : ℕ) : W.Ψ (2 * (m + 3)) * W.ψ₂ =
     W.Ψ (m + 2) ^ 2 * W.Ψ (m + 3) * W.Ψ (m + 5) - W.Ψ (m + 1) * W.Ψ (m + 3) * W.Ψ (m + 4) ^ 2 := by
-  repeat erw [Ψ_ofNat]
+  repeat rw_mod_cast [Ψ_ofNat]
   simp_rw [preΨ'_even, if_pos <| even_two_mul _, Nat.even_add_one, ite_not]
   split_ifs <;> C_simp <;> ring1
```

Comment:
```suggestion
  split_ifs <;> C_simp <;> ring
```
Can you check the other occurrences of `ring1` and leave `ring`? The first is useless here and a bit too strong for what is needed.

Reply:
I don't understand what you mean by `ring1` being useless/strong here. I believe `ring` is just `ring1 <|> ring_nf`, and I always prefer `ring1` because I *know* I won't need to get to `ring_nf`. There are examples of goals (which I cannot replicate right now) that are solved by `ring_nf` but not by `ring1`. I think something like this is volatile code, so closing goals with `ring1` means that I will never need to consider this scenario (i.e. it closes the goal iff it is solved by both `ring1` and `ring_nf`). In any case this is a thing I did in all the elliptic curve files, so this refactor should be done globally if at all, and hence would be outside the scope of this feat PR.

This applies to your other `ring`-related comments too.

Reply:
OK. I see what you mean , no problem.